### DISCUSSION TITLE
Open3e und Node-Red

### DISCUSSION THREAD

**Discussion by neandr** [open3e/open3e#148] [Ideas]
Neben ioBroker oder HomeAssistant sollte ein einfaches Setup für die Verbindung von Open3E zu Node-Red unterstützt sein.

Grundsätzlich sind wohl alle Komponenten vorhanden, um z.B. eine Viessmann Installation (VX3, Vcal, E380) mit CAN-Bus auf einem Raspberry (RPI) zu unterstützen:

    -- Vi CAN-Bus External  <==> Can-Bus Adapter <==> per USB am Raspberry angeschlossen
    -- Raspberry mit MQTT Broker, Open3e und Node-Red

So ist nach Installation von Open3e und dem initialen Setup der Listen Mode an der RPI Konsole verfügbar mit, z.B.:
> ~ $ open3e -c can0 -dev vcal -m 192.168.178.70:1883:open3e -mfstr "{didNumber}_{didName}" -l open3e/cmnd
> Enter listener mode, waiting for commands on mqtt...

Damit ist Node-Red /MQTT bereit mit der Vi Anlage zu arbeiten. Um die anderen VI-Komponenten anzusprechen, sind zusätzliche Aufrufe mit dem `-dev` Parameter notwendig.

Ein einfaches Node-Red Beispiel zur Abfrage von zwei Datenpunkten:
![image](https://github.com/user-attachments/assets/3ab9150b-9448-402a-85dc-5dc86f0cdf88)

Der dump Node zeigt dann u.a. 
```
11/8/2024, 7:18:49 PM[node: received E3 data](http://raspy:1880/#)open3e/cmnd : msg.payload : Object
{ mode: "read", data: array[2] }
11/8/2024, 7:18:49 PM[node: received E3 data](http://raspy:1880/#)open3e/271_DomesticHotWaterSensor/Actual : msg.payload : number
41
11/8/2024, 7:18:49 PM[node: received E3 data](http://raspy:1880/#)
open3e/271_DomesticHotWaterSensor/Minimum : msg.payload : number
35.6
```

Dieses Setup hat einen Schönheitsfehler! Damit Node-Red den Datenaustausch machen kann, werden  Konsolbefehle abgesetzt, das ist sicher keine "normale" Betriebsart. Beim Schliessen des Terminals bricht die Kommunikation ab.

Wünschenswert wäre es, Open3e für die Systemkomponenten so zu starten, dass sie für die installierten VI-Komponenten mit System-boot scharf gemacht werden.
Wäre ein Setup mit `systemctl status|start|stop` möglich oder sollte `open3e` zusätzliche Modi bekommen?

**Comment by MyHomeMyData** [comment]
Wie man open3e als Service einrichtet ist [hier](https://github.com/open3e/open3e/wiki/030-Installation-und-Inbetriebnahme-von-open3E#open3e-als-service-einrichten-und-bei-systemstart-automatisch-starten) beschrieben.

Ich kenne mich mit Node-Red nicht aus. Prinzipiell sollte sich die Kommunikation über MQTT aber per Listener-Mode abwickeln lassen. Falls dafür bestimmte Formate notwendig sind, können wir das gerne diskutieren.

Gruß, Jürgen

**Comment by neandr** [comment]
Danke für den Hinweis auf die Service Einrichtung.
In `ExecStart=/bin/bash -c 'cd /home/pi/open3e && /home/pi/open3e/.venv/bin/open3e @args'` 
wird open3e als ein Prozess gestartet mit den Parametern in@args.
Damit läuuft eine Instanz von open3e. Aber der open3e Befehl verträgt nur einen -dev Parameter. So müssten für jedes E3 Gerät ein anderer/zusätzlicher Service mit anderem Namen eingerichtet werden? Also in Bereich
```
[Unit]
Description=Open3e Service Script
``` 
dann `Description=Open3eVX3 Service Script `   oder  `Description=Open3eE380 Service Script`
Richtig? Und das läuft dann für Open3e ohne Probleme?

**Reply by MyHomeMyData** [reply]
Nein, das funktioniert über einen Scan der Systemumgebung. Anschließend übergibt man open3e die Device-Liste (devices.json). Ist im gleichen Wiki-Artikel beschrieben, wie die Service-Einrichtung.

**Comment by flecke-m** [comment]
Alternativ gibt es open3e auch als Docker Container, hier kann man mit "restart always" auch eine dauerhafte Funktion (auch nach Systemstart) sicherstellen.

**Comment by neandr** [comment]
@flecke-m
"Docker" das unbekannte Etwas ... habe mich (bisher) noch nie damit befasst. Sollte ich?

**Reply by flecke-m** [reply]
Nein muss man nicht. Wenn man mehrere Services auf einem Raspberry oder PC laufen, hat man eine bessere Trennung. Ich habe z.B. open3e und e3oncan im Docker Container laufen um VX3 und das Energy Meter aus zu lesen.

**Reply by MyHomeMyData** [reply]
Das macht natürlich Sinn, denn open3e und e3oncan decken verschiedene Bereiche ab und können sich nicht gegenseitig stören.

Betreibt man mehrere open3e Instanzen parallel, hat man das Risiko von Störungen der CAN-Kommunikationen.

**Comment by neandr** [comment]
@MyHomeMyData 
> _Nein_, das funktioniert über einen Scan der Systemumgebung. Anschließend übergibt man open3e die Device-Liste (devices.json). Ist im gleichen Wiki-Artikel beschrieben, wie die Service-Einrichtung

Worauf bezieht sich das "_Nein_"? 
Und würde ich drei verschiedene Services für die drei Geräte (VX, Vcal, E380) einrichten müssen? Oder gibt es da andere Befehle oder ggf. Erweiterung des `open3e` Programms, bzw. wäre das sinnvoller?

**Reply by MyHomeMyData** [reply]
> dann Description=Open3eVX3 Service Script  oder Description=Open3eE380 Service Script
Richtig? Und das läuft dann für Open3e ohne Probleme?

Das Nein bezieht sich auf Deine Frage. Lies doch bitte den [Wiki-Artikel](https://github.com/open3e/open3e/wiki/030-Installation-und-Inbetriebnahme-von-open3E#scannen-der-systems), das sollte Deine Fragen beantworten. Vom Start mehrere Services rate ich ab. Das macht die Sache unnötig kompliziert und birgt die Gefahr von Konflikten auf dem CAN-Bus, falls mehrere open3e-Instanzen auf der selben Device-Adresse operieren.

**Comment by neandr** [comment]
@MyHomeMyData 
> Das Nein bezieht sich auf Deine Frage. Lies doch bitte den [Wiki-Artikel](https://github.com/open3e/open3e/wiki/030-Installation-und-Inbetriebnahme-von-open3E#scannen-der-systems), das sollte Deine Fragen beantworten. 

Bitte kläre mich auf, wenn meine folgenden Überlegungen falsch sind:

Der Wiki-Artikel bringt mir leider nicht die Klarheit, sorry, er geht nicht auf die grundsätzlichen Fragen ein.

1.  die (derzeit) 3 Systeme (VX, VCAL, E380) lassen sich nicht mit <ins>einem</ins> `open3e` Befehl für den Listen Mode aktivieren; für jedes `-dev` Gerät muss ein entsprechender Befehl die Kommunikation initiieren, siehe wie in meinem ersten Beitrag oben.

> Vom Start mehrere Services rate ich ab. Das macht die Sache unnötig kompliziert und birgt die Gefahr von Konflikten auf dem CAN-Bus, falls mehrere open3e-Instanzen auf der selben Device-Adresse operieren.

2. Könnten mehrere Instanzen von `open3e` aufgesetzt werden, die jeweils <ins>ein</ins>  `-dev` Gerät adressieren? Mit Service `open3e.service` lässt sich das wohl nicht machen, auch wenn jede Instanz nur ein Gerät adressiert. Würde Docker hier die notwendige Separierung bringen? Und wäre das dann wirklich konfliktfrei für den CAN-Bus?  
3. Den` ioBroker.e3oncan` verstehe ich so, dass Änderungen der Datenpunkte aller E3 Geräte laufend erfasst, mit der Liste der Datenpunkte in json formatiert werden und dann in ioBroker zur Verfügung stehen.
  a) Dieser Mode wäre auch aus Sicht von Node-Red die einfachste Lösung, um die Datenpunkte auszulesen/weiter zu verarbeiten. Besteht die Möglichkeit open3e dahingehend einzusetzen / zu erweitern?      
  b) Mit welchem Programm / Befehl würde dann das Schreiben von Datenpunkten parallel dazu möglich sein?

**Reply by MyHomeMyData** [reply]
Der Schlüssel zur Adressierung ist `open3e_depictSystem`. Es wird die devices.json erzeugt und dann ist "komplexe Adressierung" möglich. Zitat aus dem Readme von open3e:

> To refer to a specific device, you have to use "complex addressing mode", e.g. to read dids 268 and 269 of device 0x680 use -cnfg devices.json -r 0x680.268,0x680.269. Do not use option -dev in this context.

> It's possible to change naming of devices by editing devices.json: When you change "0x680": { to "vitocal": { you can use read commands like -r vitocal.268,vitocal.269. New naming will also be used for paramter {device}within MQTT topics (see below).

Du kannst auch mehrere Services aufsetzen und diese über -dev verschiedenen Devices, also Adressen, zuweisen. Die Gefahr dabei ist, dass man z.B. nach einem Vierteljahr nochmal was erweitert und einen weiteren Service auf einer bereits verwendeten Adresse ergänzt. Dann werden seltsame Fehler auftreten und die open3e-Instanzen mit Fehlermeldungen beendet werden. Wahrscheinlich nur sporadisch. Da hat man dann viel Spaß bei der Fehlersuche. Die Probleme entstehen dabei immer auf dem CAN-Bus, wenn mehrere Instanzen Kommunikationen zum gleichen Device aufbauen, die sich zeitlich überlagern.

Wenn Du ioBroker benutzt, würde ich aber ohnehin den ioBroker.e3oncan Adapter verwenden. **Dann brauchst Du open3e gar nicht.**

Der Adapter sorgt dafür, dass keine Bus-Konflikte entstehen und unterstützt auch das Energy Meter E380. Nach der Installation solltest Du unbedingt einen Device-Scan und einen Datenpunkt-Scan durchführen!

Es werden zwei Modi unterstützt:
* UDSonCAN: Das entspricht der Funktionalität von open3e, nutzt dazu aber nicht open3e. Du definierst in der Adapter-Konfiguration für jedes Gerät Datenpunktlisten und Zeitspannen zur aktiven Abfrage der Datenpunkte. Schreiben ist ganz einfach möglich, indem der zugehörige Objekt-State in ioBroker geändert wird (mit Ack=false). Der Adapter löst dann einen Schreibvorgang aus und sorgt dafür, dass nicht gleichzeitig auf dem gleichen Device gelesen wird. Nach 2,5 Sekunden wird der Datenpunkt automatisch wieder gelesen. So sieht man direkt, ob der Schreibvorgang funktioniert hat.
* Collect: Hier wird auf dem CAN-Bus nur gelauscht, also **nicht** aktiv Daten angefordert. Das funktioniert beim Energy Meter und bei E3-Geräten, sofern sie vernetzt sind. Dazu gibt's auch einen [Artikel](https://github.com/MyHomeMyData/ioBroker.e3oncan/discussions/34).

Aus beiden Quellen wird eine gemeinsame Objekt-Datenstruktur erzeugt, die alle Datenpunkte enthält. Jeder Datenpunkt wird parallel in 3 Formaten bereitgestellt: Als Raw-Daten (so wie das Device die Daten anliefert), als JSON (falls wir die Struktur kennen) und als TREE (JSON heruntergebrochen auf einzelnen Elemente).

**Reply by philippoo66** [reply]
da ich schon länger mit @neandr zugange bin, mische ich mich mal kurz ein, falls ihr erlaubt... @MyHomeMyData bitte korrigieren mich, wenn ich was falsches sage!

1. sowohl der ioBroker.e3oncan Adapter im USDonCAN Modus als auch open3e senden aktiv Requests auf dem CAN Bus und empfangen entsprechenden die Responses. Deswegen kann (sicher) **nur eins von beidem und das auch nur in einer Instanz betrieben** werden! Ansonsten besteht die Gefahr, dass da was durcheinander läuft (was nicht zwingend zum Abstruz führen muss aber das Ergebnis ist nicht vorherzusagen)
2. sowohl der ioBroker.e3oncan Adapter im Collect Modus als auch e3onCan (das Python Projekt) lauschn nur dem Datenverkehr auf dem Bus und können entsprechend gefahhrlos neben den aktiv 'sprechenden' Sachen (siehe 1.) laufen, theoretisch auch in mehreren Instanzen, was aber wenig Sinn macht.
3. open3e: das Arbeiten mit der `-dev` Option ist nur ein 'Behelf' und sollte vermieden werden! Statt dessen depict_system durchlaufe lassen und dann open3e mit `-cnfg devices.json` verwenden ODER den ioBroker Adapter (der macht das Gleiche, nur eleganter). Hiermit sind dann **alle am CAN befindlichen Devices per UDS erreichbar**. Das Stichwort 'komplexe Adressierung' wurde schon genannt. Das Energy Meter ist kein UDS Gerät und entsprechend per UDS nicht auszulesen (und es wird zumindest in open3e auch nicht 'belauscht', open3e funktioniert ausschliesslich UDS).

Grundsätzlich würde ich empfehlen entweder ioBroker (mit dem Adapter) _oder_ NodeRed (mit open3e) zu benutzen! Zur Not (falls es unvermeidlich ist, beide HAs zu benutzen) sollte man "eine Brücke" per MQTT zwischen NodeRed und ioBroker bauen. Es ist aber dringend zu vermeiden, einen Konflikt wie unter 1. beschrieben zu provozieren.

Wenn man Datenpunkte schreiben oder aktiv requesten will, tut man das im produktiven Betrieb per Listener Mode (mit open3e) oder per dem Pendant des ioBroker Adapters. 

Grundsätzlich ist der ioBroker Adapter weiterentwickelter und komfortabler in der Benutzung als open3e. Es gibt glaubich nichts, was mit open3e ginge und mit dem Adapter nicht. Deswegen sollte die Wahl des Mittels klar sein, wenn man ioBroker benutzt.

Es sollte kein Problem sein, die ioBroker Adapter Daten (z.B. per Skript?) auch nach MQTT zu publishen und so in NodeRed verfügbar zu haben. Eine 'Empfangsmöglichhkeit' für aus NodeRed gesendete Requests besteht glaubich(?) mit dem ioBroker Adapter nicht 'von der Stange', lässt sich aber bestimmt auch realisieren.

Grundsätzlich (zum 3. Mal) gilt: sauber bleiben und und nichts von hinten durch die Brust in's Auge vermischen!

**Reply by MyHomeMyData** [reply]
Hallo Phil, ja, denke das passt alles. E3onCAN (Python) kann man auch mehrfach starten und auf verschiedene Adressen, also Devices, lauschen, z.B. auf Vitocal und VX3 und E380. Da kann es keine Konflikte geben, da ja auf dem CAN-Bus nur zugehört wird.

Der ioBroker.e3oncan Adapter kann auch parallel auf mehreren Adressen lauschen. Man kann das in der Adapter-Konfiguration auswählen.
Die Daten lassen sich ganz einfach per MQTT weiterleiten. Man muss lediglich den MQTT-Adapter installieren und kann dann für jedes Objekt konfigurieren, wie es per MQTT versendet werden soll. Natürlich auch für jeden einzelnen Datenpunkt der e3oncan Adapters. Ein Skript ist dazu nicht notwendig.
Ebenso kann man damit MQTT-Messages empfangen, z.B. von Node-Red. Es sollte also auch möglich sein, das so zu konfigurieren, dass ein e3oncan-Datenpunkt von Node-Red per MQTT empfangen wird und direkt ins E3-Device geschrieben wird. Ausprobiert habe ich das aber nicht.

**Comment by neandr** [comment]
@MyHomeMyData 
Zu diesen Beitrag
> Wenn Du -dev vx3 angiebst, werden nur die Daten der VX3 dekodiert. Ohne Angabe von -dev wird alles dekodiert, was dekodierbar ist.

war da noch eine Korrektur, aber es sieht wohl etwas anders aus, wenn `-dev` weggelassen wird:

`~/open3e $ python3 ./Open3Eclient.py -c can0 -m 192.168.178.70:1883:open3e/vxyz -mfstr "{didNumber}_{didName}" -l open3e/cmnd -v`

Dieser console Befehl zeigt 
a) an der console die Daten für die MQTT Antwort (hier nur für 274)
```
None 271 DomesticHotWaterSensor {"Actual": 41.4, "Minimum": 35.1, "Maximum": 53.7, "Average": 41.4, "Unknown": 0}
None 274 OutsideTemperatureSensor {"Actual": 8.6, "Minimum": 1.9, "Maximum": 15.1, "Average": 8.2, "Unknown": 0}
```
b) in z.B. Node-Red das MQTT Ergebnis
```
node: received E3 data
open3e/vxyz/274_OutsideTemperatureSensor/Average : msg : Object
   object
      topic: "open3e/vxyz/274_OutsideTemperatureSensor/Average"
      payload: 8.2
      qos: 0
      retain: false
      _msgid: "f099ea42bc7eaaaa"
```
Fehlt `-dev` im console Befehl, erscheint mit `-v`  in der Ausgabe `None` als führender Datenpunkt.

Da im Beispiel mit `-m` nicht nur `:1883:open3e` mitgegeben wird, sondern `:1883:open3e/vxyz` beinhaltet der MQTT Satz im topic dieses Element ebenso! 

Damit sind im Node-Red (und anderen MQTT Systemen) leichtere/variablere Zuordnungen der gesendeten/empfangenen Daten möglich.

**Reply by philippoo66** [reply]
das `None` kommt in der Konsole, weil mit `-v` immer die komplette Informations'palette' ausgegeben wird. Würdest  mit `-cnfg` arbeiten, stände da die ECU statt None. Ohne -cnfg gibt es aber keine device Liste.

**Reply by philippoo66** [reply]
by the way - wenn du Gruppierungen al la vxyz haben möchtest, kannst du das auch in der Datenpunktliste vor den Namen schreiben z.B. "vxyz/OutsideTemperatureSensor" statt "OutsideTemperatureSensor". Nur muss dann die Nummer hinter den Namen kommen: `-mfstr "{didName}_{didNumber}"`. Probelmatisch wird das allerdings, wenn es OutsideTemperatureSensor mit vxyz und vabc gibt.

Man kann auch in dem format string slashs statt Unterstriche verwenden, dann erreicht man damit auch eine weitere Ebene: `-mfstr "{device}/{didNumber}_{didName}"`
